### PR TITLE
added from list

### PR DESCRIPTION
**Pull request by abdelaziz-mahdy**
fix https://github.com/rainyl/opencv_dart/issues/164

what do you think?

### CODE REVIEW

**Comment by rainyl** [comment]
@abdelaziz-mahdy Thanks for your contribution!

An important problem is that, the multi-dimenssional List in dart is [jagged array](https://en.wikipedia.org/wiki/Jagged_array), which means users may input some 2D list without alignment.

so maybe we should warn users if the list is not aligned, e.g., 
```dart
final data = [
  [1, 2],
  [4, 5, 6, 7],
  [7, 8, 9],
];
```

File: test/core/mat_test.dart [hunk 1]
```diff
@@ -766,4 +766,44 @@ array([[[  0,   1,   2], [  3,   4,   5], [  6,   7,   8]],
...
     }
   });
+
+  test('Mat Creation from 2D List', () {
+    final data2D = [
+      [1, 2, 3],
+      [4, 5, 6],
+      [7, 8, 9]
+    ];
+    final matFrom2DList = cv.Mat.from2DList(data2D, const cv.MatType.CV_8UC(2));
```
Comment (rainyl) [inline test/core/mat_test.dart:776]: Maybe here should be `const cv.MatType.CV_8UC(1)`?

File: lib/src/core/mat.dart [hunk 2]
```diff
@@ -1318,6 +1318,30 @@ class Mat extends CvStruct<cvg.Mat> {
...
+  /// [data] should be a 3D list of numbers where each sublist represents a row of the matrix.
+  /// [type] specifies the Mat type.
+  factory Mat.from3DList(List<List<List<num>>> data, MatType type) {
+    final rows = data.length;
+    final cols = data[0].length;
+    final channels = data[0][0].length;
+    assert(channels == type.channels, "The number of channels in the data must match the Mat type.");
+    final flatData = data.expand((row) => row.expand((pixel) => pixel)).toList();
+    return Mat.fromList(rows, cols, type, flatData);
+  }
```
Comment (rainyl) [inline lib/src/core/mat.dart:1343]: Maybe place these two new constructors exactly after `Mat.fromList` is better.

**Comment by abdelaziz-mahdy** [comment]
> @abdelaziz-mahdy Thanks for your contribution!
> 
> An important problem is that, the multi-dimenssional List in dart is [jagged array](https://en.wikipedia.org/wiki/Jagged_array), which means users may input some 2D list without alignment.
> 
> so maybe we should warn users if the list is not aligned, e.g.,
> 
> ```dart
> final data = [
>   [1, 2],
>   [4, 5, 6, 7],
>   [7, 8, 9],
> ];
> ```

do you have any idea how to check on that without reducing performance? i like the idea but i cant think of a way to check on it without checking when expanding that the length is the same

**Comment by abdelaziz-mahdy** [comment]
i added checks to ensure the input is correct, to avoid bugs, let me know what you think

**Comment by rainyl** [comment]
> do you have any idea how to check on that without reducing performance? i like the idea but i cant think of a way to check on it without checking when expanding that the length is the same

Since dart have no such thing as `multidimensional array` like [C#](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/arrays), so the easiest way is to check the element length of a 2D array, i.e., `data.every((e)=>e.length==data.first.length);`, but I am not sure how much performance will be lost.

**Comment by abdelaziz-mahdy** [comment]
> > do you have any idea how to check on that without reducing performance? i like the idea but i cant think of a way to check on it without checking when expanding that the length is the same
> 
> Since dart have no such thing as `multidimensional array` like [C#](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/arrays), so the easiest way is to check the element length of a 2D array, i.e., `data.every((e)=>e.length==data.first.length);`, but I am not sure how much performance will be lost.

i did add the checks as a separated loops if you think they should be included in the expand (i think it will be better to iterate the list only once) let me know

**Comment by rainyl** [comment]
> i did add the checks as a separated loops if you think they should be included in the expand (i think it will be better to iterate the list only once) let me know

Agree that it should be iterated once, I changed it to `linq` style to make it clearer and more concise, hope you won't mind it.

**Comment by rainyl** [comment]
@abdelaziz-mahdy Looks good to me now, what do you think? If no further problems I am going to merge it.

**Comment by abdelaziz-mahdy** [comment]
looks good to me, but the every keyword will make the loop iterated first then flattened so the loop is iterated twice? also asserts or exception? asserts only works in debug right? that will lead to crashes on release this is why i used exception?

**Comment by rainyl** [comment]
> looks good to me, but the `every` keyword will make the loop iterated first then flattened so the list is iterated twice? also asserts or exception? asserts only works in debug right? that will lead to crashes on release this is why i used exception?

Oh, you are right, I didn't know the assert in dart will only throw in debug mode, such a strange feature....

I have fixed it, could you please review it?

**Comment by rainyl** [comment]
Merged, thanks for your contribution!

**Pull request merged**